src/puppetlabs/puppetdb/command.clj

```diff
@@ -296,25 +296,39 @@
 
 ;; Fact replacement
 
+(defn remove-facts-regex [facts-blacklist fact-map]
+  ;; TODO this checks even when it finds a match, is there a better way to write this?
+  (->> (for [patt facts-blacklist
+             kv fact-map
+             :let [k (first kv)]
+             :when (re-matches patt k)]
+         k)
+       (apply dissoc fact-map)))
+
 (defn replace-facts*
-  [{:keys [payload id received] :as command} start-time db facts-blacklist]
+  [{:keys [payload id received] :as command} start-time db
+   {:keys [facts-blacklist facts-blacklist-type]}]
   (let [{:keys [certname package_inventory] :as fact-data} payload
         producer-timestamp (:producer_timestamp fact-data)
+        remove-blacklist-facts (if (= "regex" facts-blacklist-type)
+                                 (partial remove-facts-regex facts-blacklist)
+                                 ;; if blacklist-type is not "regex" assume it is "regular"
+                                 #(apply dissoc % facts-blacklist))
         trimmed-facts (cond-> fact-data
-                        (seq facts-blacklist) (update :values #(apply dissoc % facts-blacklist))
+                        (seq facts-blacklist) (update :values remove-blacklist-facts)
                         (seq package_inventory) (update :package_inventory distinct))]
     (jdbc/with-transacted-connection' db :repeatable-read
       (scf-storage/maybe-activate-node! certname producer-timestamp)
       (scf-storage/replace-facts! trimmed-facts))
     (log-command-processed-messsage id received start-time :replace-facts certname)))
 
 (defn replace-facts
-  [{:keys [payload version received] :as command} start-time db facts-blacklist]
+  [{:keys [payload version received] :as command} start-time db fact-blist-map]
```

Comment:
Maybe this should be named something like `facts-blacklist-config`. I know it's wordy but bits are cheap and it may help prevent some confusion. There are numerous usages below as well, if we decide we want to change it, we should update those as well.